### DISCUSSION TITLE
Wrong flash.xml.in file deployed to tegraflash with mender

### DISCUSSION THREAD

**Discussion by brgl** [OE4T/tegra-demo-distro#41] [Q&A] [answered]
Hi @madisongh 

Thank you for your work on the tegra support in yocto!

I'm trying to add mender support to my custom distro for Xavier NX and have hit a problem I'm not sure how to solve. I'm aware that the issue is in my code because the demo works fine but I thought I'd ask here for advice anyway.

I'm basing my work on the `tegrademo-mender` distro except that I moved the Xavier NX-specific variable assignments to my machine conf file. Everything builds fine but when I run the `dosdcard.sh` script, I'm hitting the following error in `nvflashxmlparse`:

```
Traceback (most recent call last):
  File "/tmp/tegra/nvflashxmlparse", line 157, in <module>
    ret = main()
  File "/tmp/tegra/nvflashxmlparse", line 131, in main
    ', '.join(list(layout.devices.keys()))))
RuntimeError: Device type 'sdcard' not present; available types: sdmmc_boot, sdmmc_user
No partition definitions found in signed/flash.xml.tmp

```
After some digging I noticed that the demo deploys a different `flash.xml.in` file into the tarball and that the file I have in my build does not indeed have an sdcard section. The file I'm seeing seems to be based on the one in `meta-mender-tegra` under `recipes-bsp/tegra-binaries/mender-custom-flash-layout/tegra194/flash_mender.xml` while the one used by demo looks more like a modified version of the one generated by `meta-tegra`.

My image classes used for building:

```
IMAGE_CLASSES=" image_types_tegra mender-part-images mender-ubimg mender-artifactimg mender-dataimg mender-bootimg mender-datatar image_types_mender_tegra license_image"
```

These are exactly the same as in the demo image.

Do you have any hints as to what the culprit may be?

**Comment by dwalkes** [comment, marked as answer]
@brgl which branch are you using?  My first guess was that you aren't using the [dynamic layer](https://github.com/OE4T/tegra-demo-distro/tree/master/layers/meta-tegra-support/dynamic-layers/meta-mender-tegra/recipes-bsp/tegra-binaries/mender-custom-flash-layout) files somehow but I don't see any tegra194 layouts there so I don't know that this would explain it.

**Reply by brgl** [reply]
I'm on `dunfell` everywhere (`dunfell-l4t-r32.4.3` on `meta-tegra`). This is not the first time I'm seeing the expression "dynamic layers" but I'm still a bit confused as to how they work and what is their purpose. Does the demo use them? For the demo build I'm just doing `source ./setup-env --distro tegrademo-mender --machine jetson-xavier-nx-devkit` so nothing unusual.

In the case of the demo build: which package generates the .xml with layout exactly?

**Reply by madisongh** [reply]
For tegrademo-mender builds, the XML file is provided by the `mender-custom-flash-layout` recipe, from the `meta-mender-tegra` layer (on dunfell-l4t-r32.4.3, at least).  From the error message, it definitely looks like you've got the wrong flash layout (maybe `jetson-xavier-nx-devkit-emmc`'s ?).  You should compare it against the copy at `layers/meta-mender-tegra/recipes-bsp/tegra-binaries/mender-custom-flash-layout/jetson-xavier-nx-devkit/flash_mender.xml` - that's the one that should be getting picked up.

As for "dynamic" layers, they're a way of conditionally including some recipes/bbappends based on whether a layer is present in the build.  There's a brief description in [the Yocto documentation](https://docs.yoctoproject.org/singleindex.html#term-BBFILES_DYNAMIC).

**Reply by brgl** [reply]
So indeed it seems like the layout xml is being fetched from `tegra194` sub-directory of `mender-custom-flash-layout` and not `jetson-xavier-nx-devkit`. Looking at the recipe, I don't see any obvious way this directory is chosen so I assume these are some bitbake internals based on the `MACHINE` variable.

I was using a custom machine config file (including the original Xavier NX config) with a different name so this is what probably caused it. If I switch my `MACHINE` to `jetson-xavier-nx-devkit` then the flash xml is picked correctly. If I want to have a custom machine file - I guess I'll need to copy the xavier xml file into my own layer and recreate a similar directory structure `mender-custom-flash-layout` uses? Is there any other, more correct approach?

**Reply by madisongh** [reply]
That makes sense, then.  Yes, you'll need a bbappend in your layer to supply the correct flash layout file for your custom machine. The value of MACHINE is automatically included as a subdirectory in the FILESPATH variable that is used for locating `file://` files in the SRC_URI.  (Other variables are also treated this way.)

**Reply by brgl** [reply]
Ok, now everything works correctly. Thanks a lot for the support.

**Reply by dwalkes** [reply]
> If I want to have a custom machine file - I guess I'll need to copy the xavier xml file into my own layer and recreate a similar directory structure mender-custom-flash-layout uses? Is there any other, more correct approach?

This thread got buried in my inbox and I'm getting back to this late but I wanted to mention we have a page at https://github.com/OE4T/meta-tegra/wiki/Creating-a-custom-MACHINE#custom-machine-definitions-for-existing-hardware describing how to custoimze the MACHINE variable.  If there's anything you've learned here which would be useful to add to that page any updates would be great.  Glad you've got everything working!

**Reply by brgl** [reply]
@dwalkes Thanks, I eventually ended up just using machine overrides for xavier nx in my distro file but I'll use this when we switch to the custom carrier board design.
